lib/learn_open/environments/base_environment.rb

```diff
@@ -69,11 +69,31 @@ def notify_of_completion
       def warn_if_necessary(lesson)
         return unless lesson.later_lesson
 
+        warn_response = nil
+        prompt_count = 0
+        # implementation uses chomp, thus a valid \n becomes ''
+        post_chomped_enter_key = ""
+        valid_responses = ['yes', 'y', 'no', 'n', post_chomped_enter_key]
+        visible = -> (token){ token.match(/\w+/) }
+        valid_resp_insert = [
+          *valid_responses.select{ |c| visible.call(c) },
+          "<ENTER>"
+        ].join(', ')
+
         io.puts 'WARNING: You are attempting to open a lesson that is beyond your current lesson.'
         io.print 'Are you sure you want to continue? [Y/n]: '
 
-        warn_response = io.gets.chomp.downcase
-        exit if !['yes', 'y'].include?(warn_response)
+        until valid_responses.include?(warn_response) or
+              prompt_count > 2
+          warn_response = io.gets.chomp.downcase
+          prompt_count += 1
+        end
+
+        if prompt_count > 2
+          raise IOError, "Valid inputs are: #{valid_resp_insert}"
```

Comment:
Perhaps we should just loop until we get one of the 4 valid responses? I think raising an error in the in-browser IDE will give a non ideal experience to the student.

Reply:
@vaughanj10 

Removing this as a logic creates a challenge for the IO mocking. Specifically, it mock will return a bad input for the multiple calls. Raising the exception allows us to see that the loop state fired to exhaustion and raised an error. Maybe it should be after at least some higher max_retries value.

This might be a limit of my capability of testing, though.

OTOH, we _could_ simply drop the tests as well as the prompt_count testing logic. IMHO we went from failing silently, to failing informatively, a step forward.